framework/Optimizers/RavenSampled.py

```diff
@@ -0,0 +1,566 @@
+    """
+      Provides the next sample to take.
+      After this method is called, the self.inputInfo should be ready to be sent
+      to the model
+      @ In, model, model instance, an instance of a model
+      @ In, inp, list, a list of the original needed inputs for the model (e.g. list of files, etc.)
+      @ Out, None
+    """
+    # get point from stack
+    point, info = self._submissionQueue.popleft()
+    point = self.denormalizeData(point)
+    # assign a tracking prefix
+    prefix = self.inputInfo['prefix']
+    # register the point tracking information
+    self._registerSample(prefix, info)
+    # build the point in the way the Sampler expects
+    for var in self.toBeSampled: #, val in point.items():
+      val = point[var]
+      self.values[var] = val # TODO should be np.atleast_1d?
+      ptProb = self.distDict[var].pdf(val)
+      # sampler-required meta information # TODO should we not require this?
+      self.inputInfo['ProbabilityWeight-{}'.format(var)] = ptProb
+      self.inputInfo['SampledVarsPb'][var] = ptProb
+    self.inputInfo['ProbabilityWeight'] = 1 # TODO assume all weight 1? Not well-distributed samples
+    self.inputInfo['PointProbability'] = np.prod([x for x in self.inputInfo['SampledVarsPb'].values()])
+    self.inputInfo['SamplerType'] = self.type
+
+  def localFinalizeActualSampling(self, job, model, inp):
+    """
+      Runs after each sample is collected from the JobHandler.
+      @ In, job, Runner instance, job runner entity
+      @ In, model, Model instance, RAVEN model that was run
+      @ In, inp, list, generated inputs for run
+      @ Out, None
+    """
+    Optimizer.localFinalizeActualSampling(self, job, model, inp)
+    # TODO should this be an Optimizer class action instead of Sampled?
+    # collect finished job
+    prefix = job.getMetadata()['prefix']
+    # If we're not looking for the prefix, don't bother with using it
+    ## this usually happens if we've cancelled the run but it's already done
+    if not self.stillLookingForPrefix(prefix):
+      return
+    if job.getReturnCode() != 0: # TODO shouldn't this be "if job failed"?
+      raise NotImplementedError # FIXME   handle failed runs
+    # FIXME implicit constraints probable should be handled here too
+    # get information and realization, and update trajectories
+    info = self.getIdentifierFromPrefix(prefix, pop=True)
+    _, full = self._targetEvaluation.realization(matchDict={'prefix': prefix})
+    # trim down opt point to the useful parts
+    # TODO making a new dict might be costly, maybe worth just passing whole point?
+    ## testing suggests no big deal on smaller problem
+    rlz = dict((var, full[var]) for var in (list(self.toBeSampled.keys()) + [self._objectiveVar] + list(self.dependentSample.keys())))
+    # the sign of the objective function is flipped in case we do maximization
+    # so get the correct-signed value into the realization
+    if self._minMax == 'max':
```

Comment:
this operation should be clearly reported in the docstirng of the class... Just for clarity.

Reply:
added to docstrings